### ISSUE TITLE
Accuracy and IoU Different on Each Run

### ISSUE THREAD

**Issue by rubeea** [open-mmlab/mmsegmentation#312]
Why is my accuracy and IoU coming out to be different on each run of the experiment even if I dont change any experiment parameter. The config file is posted here for reference:

norm_cfg = dict(type='BN', requires_grad=True)
model = dict(
    type='EncoderDecoder',
    pretrained='open-mmlab://resnet50_v1c',
    backbone=dict(
        type='ResNetV1c',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        dilations=(1, 1, 2, 4),
        strides=(1, 2, 1, 1),
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=False,
        style='pytorch',
        contract_dilation=True),
    decode_head=dict(
        type='PSPHead',
        in_channels=2048,
        in_index=3,
        channels=512,
        pool_scales=(1, 2, 3, 6),
        dropout_ratio=0.1,
        num_classes=2,
        norm_cfg=dict(type='BN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss',
            use_sigmoid=False,
            loss_weight=1.0,
            class_weight=[0.02, 0.98])),
    auxiliary_head=dict(
        type='FCNHead',
        in_channels=1024,
        in_index=2,
        channels=256,
        num_convs=1,
        concat_input=False,
        dropout_ratio=0.1,
        num_classes=2,
        norm_cfg=dict(type='BN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss',
            use_sigmoid=False,
            loss_weight=0.4,
            class_weight=[0.02, 0.98])))
train_cfg = dict()
test_cfg = dict(mode='whole')
dataset_type = 'PLDUDataset'
data_root = '/content/pldu_mmsegmentation/data/pldu/'
img_norm_cfg = dict(
    mean=[104.899, 104.413, 96.448], std=[39.304, 38.998, 38.55], to_rgb=True)
crop_size = (256, 256)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations'),
    dict(type='Resize', img_scale=(360, 540), ratio_range=(0.5, 1.5)),
    dict(type='Resize', img_scale=(540, 360), ratio_range=(0.5, 1.5)),
    dict(
        type='RandomRotate',
        prob=0.5,
        degree=(45.0, 315.0),
        pad_val=0,
        seg_pad_val=255),
    dict(type='RandomCrop', crop_size=(256, 256), cat_max_ratio=0.75),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='PhotoMetricDistortion'),
    dict(
        type='Normalize',
        mean=[104.899, 104.413, 96.448],
        std=[39.304, 38.998, 38.55],
        to_rgb=True),
    dict(type='Pad', size=(256, 256), pad_val=0, seg_pad_val=255),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=[(360, 540), (540, 360)],
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[104.899, 104.413, 96.448],
                std=[39.304, 38.998, 38.55],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
data = dict(
    samples_per_gpu=2,
    workers_per_gpu=2,
    train=dict(
        type='PLDUDataset',
        data_root='/content/pldu_mmsegmentation/data/pldu/',
        img_dir='img_dir/train',
        ann_dir='ann_dir/train',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations'),
            dict(type='Resize', img_scale=(360, 540), ratio_range=(0.5, 1.5)),
            dict(type='Resize', img_scale=(540, 360), ratio_range=(0.5, 1.5)),
            dict(
                type='RandomRotate',
                prob=0.5,
                degree=(45.0, 315.0),
                pad_val=0,
                seg_pad_val=255),
            dict(type='RandomCrop', crop_size=(256, 256), cat_max_ratio=0.75),
            dict(type='RandomFlip', flip_ratio=0.5),
            dict(type='PhotoMetricDistortion'),
            dict(
                type='Normalize',
                mean=[104.899, 104.413, 96.448],
                std=[39.304, 38.998, 38.55],
                to_rgb=True),
            dict(type='Pad', size=(256, 256), pad_val=0, seg_pad_val=255),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img', 'gt_semantic_seg'])
        ],
        split=None),
    val=dict(
        type='PLDUDataset',
        data_root='/content/pldu_mmsegmentation/data/pldu/',
        img_dir='img_dir/val',
        ann_dir='ann_dir/val',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=[(360, 540), (540, 360)],
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[104.899, 104.413, 96.448],
                        std=[39.304, 38.998, 38.55],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ],
        split=None),
    test=dict(
        type='PLDUDataset',
        data_root='/content/pldu_mmsegmentation/data/pldu/',
        img_dir='img_dir/val',
        ann_dir='ann_dir/val',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=[(360, 540), (540, 360)],
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[104.899, 104.413, 96.448],
                        std=[39.304, 38.998, 38.55],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ],
        split=None))
log_config = dict(
    interval=50, hooks=[dict(type='TextLoggerHook', by_epoch=False)])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
cudnn_benchmark = True
optimizer = dict(type='Adam', lr=1e-06, weight_decay=0.0005)
optimizer_config = dict()
lr_config = dict(policy='poly', power=0.9, min_lr=0.001, by_epoch=False)
runner = dict(type='IterBasedRunner', max_iters=200)
checkpoint_config = dict(by_epoch=False, interval=200)
evaluation = dict(interval=200, metric='mIoU')
work_dir = './work_dirs/tutorial'
total_iters = 200
seed = 0
gpu_ids = range(0, 1)

Kindly help.

**Comment by xvjiarui** [comment]
You may pass `--seed 0` to reduce the variance of your experiments. However, it could not be exactly reproduced as stated [here](https://pytorch.org/docs/stable/notes/randomness.html).

**Comment by rubeea** [comment]
> You may pass `--seed 0` to reduce the variance of your experiments. However, it could not be exactly reproduced as stated [here](https://pytorch.org/docs/stable/notes/randomness.html).

I have already set it using:
cfg.seed = 0
set_random_seed(0, deterministic=False)

Should I turn on the deterministic behavior as well using torch.set_deterministic()?

Thanks

**Comment by xvjiarui** [comment]
It will be much more slow. I haven't tried.

**Comment by rubeea** [comment]
> It will be much more slow. I haven't tried.

Ok thanks for your reply. I might try it and see the difference.
